networkx/convert_matrix.py

```diff
@@ -1360,3 +1364,192 @@ def from_numpy_array(A, parallel_edges=False, create_using=None):
+    set up the function to accept the dictionary,
+    and use ``functools.partial``
+    to "reduce" the function signature to just ``(n, d)``.
+    An example below:
+
+    .. code-block:: python
+
+        from functools import partial
+        def get_molweight(n, d, mw_dict):
+            return pd.Series({"mw": mw_dict[d["amino_acid"]]}, name=n)
+
+        mw_dict = {"PHE": 165, "GLY": 75, ...}
+        get_molweight_func = partial(get_molweight, mw_dict=mw_dict)
+
+        generate_feature_dataframe(G, [get_molweight_func])
+
+    Lambda expressions are also valid!
+
+    .. code-block:: python
+
+        generate_feature_dataframe(G, lambda n, d: mw_dict.get(d["amino_acid"]))
+
+    The `name=n` piece is important,
+    as this is the piece that allows
+    the node ``n`` in the graph to become
+    the row index in the resulting dataframe.
+
+    The series that is returned from each function
+    need not only contain one key-value pair.
+    You can have two or more, and that's completely fine;
+    each key becomes a column in the resulting dataframe.
+
+    A key design choice: We default to returning DataFrames,
+    to make inspecting the data easy,
+    To pass the data into tensor libraries,
+    such as JAX, PyTorch or TensorFlow,
+    you can extract the DataFrame values after the fact.
+
+    :param G: A NetworkX-compatible graph object.
+    :param funcs: A list of functions.
+    :returns: A pandas DataFrame.
+    """
+    import pandas as pd
+
+    matrix = []
+    for n, d in G.nodes(data=True):
+        series = []
+        for func in funcs:
+            res = func(n, d)
+            if res.name != n:
+                raise NameError(
+                    f"function {func.__name__} returns a series "
+                    "that is not named after the node."
+                )
+            series.append(res)
+        matrix.append(pd.concat(series))
+
+    df = pd.DataFrame(matrix)
+    return df
+
+
+def format_adjacency(G: nx.Graph, adj, name: str):
+    """
+    Format adjacency matrix nicely.
+
+    Intended to be used when computing
+    an adjacency-like matrix off a graph object G;
+    the function will give it nice xarray names
+    that will allow indexing later on.
+
+    For example, in defining a func:
+
+    .. code-block:: python
+
+        def my_adj_matrix_func(G):
+            adj = some_adj_func(G)
+            return format_adjacency(G, adj, "xarray_coord_name")
+
+    We make some assumptions here.
+
+    1. `adj` should be a 2D matrix of shape (n_nodes, n_nodes)
+    1. `name` is something that is unique amongst all names used
+    in the final adjacency tensor.
+
+    :param G: NetworkX-compatible Graph
+    :param adj: 2D numpy array
+    :param name: A unique name for the kind of adjacency matrix
+        being constructed.
+        Gets used in xarray as a coordinate in the "name" dimension.
+    :returns: An XArray DataArray of shape (n_nodes, n_nodes, 1)
+    """
```

Comment:
Should these doc_strings be using the sphinx syntax for showing parameters in docstrings?

Reply:
Thanks for the catch, @dschult. I noticed it's NumPy-style elsewhere, so I've pushed up a change that changes the docstrings to NumPy-style.